### ISSUE TITLE
Error compiling a PBIT or PBIX from source

### ISSUE THREAD

**Issue by TonyGaul** [pbi-tools/pbi-tools#95]
Hi,
I am having an issue compiling a PBIT or PBIX from source files.  Below steps are for a PBIT, but PBIX is simialr error.

Steps:
1. Add a file called Master.pbix to repo
2. Run `pbi-tools extract '.\reports\Master.pbix'` to extract files from pbix
3. Run `pbi-tools compile '.\reports\Master' -format PBIT -outPath '.\reports\Master.pbit' -overwrite true` to compile into a PBIT
4. Code stops and shows below error
![image](https://user-images.githubusercontent.com/14303694/155652089-2bce1b2a-05d2-46f8-8a6d-ae50d6385f14.png)
5. Clicking "debug" runs a few more lines of code and the PBIT/PBIX is created.
6. I can successfully open the PBIT and load data after selecting parameters.
7. PBIX won't open and gives this error:
![image](https://user-images.githubusercontent.com/14303694/155656392-16955355-aaf1-4d80-a2a8-116e1feda6f0.png)

Here is some log info:
`pbi-tools info`
Log level: Debug
{
  "version": "1.0.0-beta.9+220202",
  "edition": "Desktop",
  "build": "1.0.22033.2227",
  "pbiBuildVersion": "2.100.1401.0",
  "amoVersion": "19.32.0.0",
  "toolPath": "C:\\Program Files\\pbi-tools\\desktop\\pbi-tools.exe",
  "settings": {
    "PBITOOLS_LogLevel": "Debug",
    "PBITOOLS_PbiInstallDir": null,
    "PBITOOLS_AppDataDir": null,
    "PBITOOLS_Debug": null
  },
  "runtime": {
    "platform": "Microsoft Windows 10.0.19044 ",
    "architecture": "X64",
    "framework": ".NET Framework 4.8.4470.0"
  },
  "pbiInstalls": [
    {
      "ProductVersion": "2.102.683.0 (22.02)",
      "Is64Bit": true,
      "InstallDir": "C:\\Program Files\\Microsoft Power BI Desktop\\bin",
      "Location": "Installer"
    }
  ],
  "effectivePbiInstallDir": "C:\\Program Files\\Microsoft Power BI Desktop\\bin",
  "pbiSessions": []
}

` pbi-tools compile '.\reports\Master' -format PBIT -outPath '.\reports\Master.pbit' -overwrite true`
Just added last chunk with relevant errors
....

File successfully opened: D:\Work\Repos\APM\powerbi\reports\Master\Model\cultures\en-US.json
Successfully deserialized part: D:\Work\Repos\APM\powerbi\reports\Master\Model
Using Power BI Desktop install: 2.102.683.0 (22.02) at C:\Program Files\Microsoft Power BI Desktop\bin
Assembly 'Microsoft.Mashup.Client.Packaging, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' found at C:\Program Files\Microsoft Power BI Desktop\bin\Microsoft.Mashup.Client.Packaging.dll  
Could not resolve assembly: Microsoft.Mashup.Client.Packaging.XmlSerializers, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
Generating PBIT file at 'D:\Work\Repos\APM\powerbi\reports\Master.pbit'...
Assembly 'Microsoft.PowerBI.Packaging, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' found at C:\Program Files\Microsoft Power BI Desktop\bin\Microsoft.PowerBI.Packaging.dll
PowerBIPackage - Intercepting: Version
ReturnValue empty: True
PowerBIPackage - Assigned value for: Version from PbiPackage
PowerBIPackage - Intercepting: DataMashup
ReturnValue empty: True
PowerBIPackage - Assigned value for: DataMashup from PbiPackage
PowerBIPackage - Intercepting: DataMashup
ReturnValue empty: True
PowerBIPackage - Assigned value for: DataMashup from PbiPackage
PowerBIPackage - Intercepting: DataModelSchema
ReturnValue empty: True
PowerBIPackage - Assigned value for: DataModelSchema from PbiPackage
PowerBIPackage - Intercepting: DataModelSchema
ReturnValue empty: True
PowerBIPackage - Assigned value for: DataModelSchema from PbiPackage
PowerBIPackage - Intercepting: DataModel
ReturnValue empty: True
PowerBIPackage - Assigned value for: DataModel from PbiPackage
PowerBIPackage - Intercepting: DataModel
ReturnValue empty: True
PowerBIPackage - Assigned value for: DataModel from PbiPackage
PowerBIPackage - Intercepting: DiagramLayout
ReturnValue empty: True
PowerBIPackage - Assigned value for: DiagramLayout from PbiPackage
PowerBIPackage - Intercepting: DiagramLayout
ReturnValue empty: True
PowerBIPackage - Assigned value for: DiagramLayout from PbiPackage
PowerBIPackage - Intercepting: ReportDocument
ReturnValue empty: True
PowerBIPackage - Assigned value for: ReportDocument from PbiPackage
PowerBIPackage - Intercepting: ReportSettings
ReturnValue empty: True
PowerBIPackage - Assigned value for: ReportSettings from PbiPackage
PowerBIPackage - Intercepting: ReportMetadata
ReturnValue empty: True
PowerBIPackage - Assigned value for: ReportMetadata from PbiPackage
PowerBIPackage - Intercepting: LinguisticSchema
ReturnValue empty: True
PowerBIPackage - Assigned value for: LinguisticSchema from PbiPackage
ReturnValue empty: True
PowerBIPackage - Assigned value for: DiagramViewState from PbiPackage
PowerBIPackage - Intercepting: Connections
ReturnValue empty: True
PowerBIPackage - Assigned value for: Connections from PbiPackage
PowerBIPackage - Intercepting: CustomVisuals
ReturnValue empty: True
PowerBIPackage - Assigned value for: CustomVisuals from PbiPackage
PowerBIPackage - Intercepting: CustomVisuals
ReturnValue empty: True
PowerBIPackage - Assigned value for: CustomVisuals from PbiPackage
PowerBIPackage - Intercepting: StaticResources
ReturnValue empty: True
PowerBIPackage - Assigned value for: StaticResources from PbiPackage
PowerBIPackage - Intercepting: StaticResources
ReturnValue empty: True
PowerBIPackage - Assigned value for: StaticResources from PbiPackage
PowerBIPackage - Intercepting: CustomProperties
ReturnValue empty: True
PowerBIPackage - Assigned empty content for: CustomProperties
PowerBIPackage - Intercepting: CustomProperties
ReturnValue empty: True
PowerBIPackage - Assigned empty content for: CustomProperties
PowerBIPackage - Intercepting: ReportMobileState
ReturnValue empty: True
PowerBIPackage - Assigned empty content for: ReportMobileState
PowerBIPackage - Intercepting: ReportMobileState
ReturnValue empty: True
PowerBIPackage - Assigned empty content for: ReportMobileState
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en-GB, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en-GB, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en, PublicKeyToken=b77a5c561934e089

#####################################################################################
Code stops here and displays error message
![image](https://user-images.githubusercontent.com/14303694/155652089-2bce1b2a-05d2-46f8-8a6d-ae50d6385f14.png)
Hitting debug runs rest of code
#####################################################################################

Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en-GB, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en-GB, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en-GB, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en-GB, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en, PublicKeyToken=b77a5c561934e089
Could not resolve assembly: mscorlib.resources, Version=4.0.0.0, Culture=en, PublicKeyToken=b77a5c561934e089
Assembly 'Microsoft.Mashup.Shims, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' found at C:\Program Files\Microsoft Power BI Desktop\bin\Microsoft.Mashup.Shims.dll

**Comment by mthierba** [comment]
Few initial questions:
- Can you reproduce the issue with the very latest pre-release? (https://github.com/pbi-tools/pbi-tools/releases/tag/1.0.0-beta.9%2B220221)
- What are the locale/language settings on your machine? Having looked around based on your screenshot, there could be an issue on machines with a non-English locale. (For better debugging support, I'll expose the active settings in `pbi-tools info` in the future.)

**Comment by mthierba** [comment]
Also, just double-checking:
- Did you [Unblock the `pbi-tools` ZIP file](https://pbi.tools/cli/#installation) before extracting it?
- Which, if any, virus scanners are running on your machine?
- Did you attempt `pbi-tools compile` using the `pbi-tools.core` version? That one has a completely different implementation, specifically one that does not rely on PBI Desktop, however, there are [no differences in the supported features](https://pbi.tools/cli/feature-matrix.html) between the two (for _compile_).

**Comment by TonyGaul** [comment]
Hi Mathias,
Thanks for the prompt reply!

I have done the following and still cannot get it to work.  I have **tried the following on both my pc and laptop** (which are pretty similar as far and OS, settings & patches go) and both have reproduced the same errors.

- Installed both win_x64 desktop and core versions (https://github.com/pbi-tools/pbi-tools/releases/tag/1.0.0-beta.9%2B220221)
- Unblocked the ZIP files for both desktop and core.
- Locale was EN-US, tested and failed.  Realised this was not my correct locale so I changed to EN-AU.  Tested and still failed.
- My local account was not an administrator, so changed to Admin.  Tested and failed.
- I am not running any virus scanners apart from Windows Defender default settings... well I haven't changed any settings
- I tried `pbi-tools extract '.\reports\Master.pbix' -modelSerialization Raw` and this produced same error
- I tried `pbi-tools.core extract '.\reports\Master.pbix'` and this failed / didn't even run.  Had same error on both computers:
PS D:\Work\Repos\APM\powerbi> `pbi-tools.core extract '.\reports\Master.pbix'`
Failure processing application bundle.
Bundle header version compatibility check failed.
A fatal error occured while processing application bundle

- So I cannot properly test the core version.  I have not set up DevOps so cannot test in a pipeline.

**Some observations:**
1. Does it matter that my .NET is 4.8.4470 and not 4.7.2?
[08:49:11.717 INF] [] Log level: Debug
{
  "version": "1.0.0-beta.9+220221",
  "edition": "Desktop",
  "build": "1.0.22052.2228",
  "pbiBuildVersion": "2.102.683.0",
  "amoVersion": "19.34.1.0",
  "toolPath": "C:\\Program Files\\pbi-tools\\desktop\\pbi-tools.exe",
  "settings": {
    "PBITOOLS_LogLevel": "Debug",
    "PBITOOLS_PbiInstallDir": null,
    "PBITOOLS_AppDataDir": null,
    "PBITOOLS_Debug": null
  },
  "runtime": {
    "platform": "Microsoft Windows 10.0.19044 ",
    "architecture": "X64",
    "framework": ".NET Framework 4.8.4470.0"
  },
  "pbiInstalls": [
    {
      "ProductVersion": "2.102.683.0 (22.02)",
      "Is64Bit": true,
      "InstallDir": "C:\\Program Files\\Microsoft Power BI Desktop\\bin",
      "SettingsDir": "C:\\Users\\tonyg\\AppData\\Local\\Microsoft\\Power BI Desktop",
      "Location": "Installer"
    }
  ],
  "effectivePbiInstallDir": "C:\\Program Files\\Microsoft Power BI Desktop\\bin",
  "pbiSessions": []
}


3. In the logs, there are so many and I am not sure how to pause them lol.  But I did manage to grab these after deleting some content in my report to reduce amount of logs.

[08:46:15.912 DBG] [PbiTools.FileSystem.ProjectFile] File not found: D:\Work\Repos\APM\powerbi\reports\Master\DiagramViewState.json
[08:46:15.913 DBG] [PbiTools.FileSystem.ProjectFile] File not found: D:\Work\Repos\APM\powerbi\reports\Master\LinguisticSchema.json
[08:46:15.914 DBG] [PbiTools.FileSystem.ProjectFile] File not found: D:\Work\Repos\APM\powerbi\reports\Master\LinguisticSchema.xml


4. In the logs I do see lots of DependenciesResolver Could not resolve assembly messages eg
[08:38:01.113 DBG] [PbiTools.Utils.DependenciesResolver] Could not resolve assembly: Microsoft.VisualStudio, PublicKeyToken=b03f5f7f11d50a3a
[08:38:02.959 DBG] [PbiTools.Utils.DependenciesResolver] Could not resolve assembly: Microsoft.AnalysisServices.Tabular.resources, Version=19.34.1.0, Culture=en-GB, PublicKeyToken=89845dcd8080cc91
[08:38:09.686 DBG] [PbiTools.Utils.DependenciesResolver] Could not resolve assembly: Microsoft.Mashup.Client.Packaging.XmlSerializers, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35

**Comment by TonyGaul** [comment]
And I tried with a completely different PBIX with different data source, visuals etc.  Same errors.

**Comment by mthierba** [comment]
Hi Tony,

> * Unblocked the ZIP files for both desktop and core.

Thanks for double-checking!

> * Locale was EN-US, tested and failed.  Realised this was not my correct locale so I changed to EN-AU.  Tested and still failed.

That's useful to know, thanks.

> * My local account was not an administrator, so changed to Admin.  Tested and failed.

Running as admin is not required for any `pbi-tools` operation.

> * I tried `pbi-tools.core extract '.\reports\Master.pbix'` and this failed / didn't even run.  Had same error on both computers:
>   PS D:\Work\Repos\APM\powerbi> `pbi-tools.core extract '.\reports\Master.pbix'`
>   Failure processing application bundle.
>   Bundle header version compatibility check failed.
>   A fatal error occured while processing application bundle
> * So I cannot properly test the core version.  I have not set up DevOps so cannot test in a pipeline.

The `extract` action is [not (yet) supported](https://pbi.tools/cli/feature-matrix.html) with the Core version. I was wondering whether you could try the `compile` step using Core - the implementation of `compile` is very different between _Core_ and _Desktop_, and since that's where your problem occurred it would be worth checking.

The _Core_ version, however, requires [the .Net 6 Runtime to be installed](https://dotnet.microsoft.com/en-us/download/dotnet/6.0/runtime). The error message you quoted seems to indicate that you don't have that.

> 
> **Some observations:**
> 
> 1. Does it matter that my .NET is 4.8.4470 and not 4.7.2?

No, doesn't matter. .Net Fx 4.7.2 is the _minimum_ version required, and 4.8 is fully supported, I opted for 4.7.2 as the minimum because that's what PBI Desktop requires, too. So I can assume that any machine running PBI Desktop already has the prerequisites for `pbi-tools` (_Desktop_) in place!  

> 2. In the logs, there are so many and I am not sure how to pause them lol.  But I did manage to grab these after deleting some content in my report to reduce amount of logs.

Yes, in _Verbose_ or _Debug_ mode, there will be loads of log messages!

However, you can redirect all console output to a file like so:

    pbi-tools.exe extract {MyReport.pbix} > output.log

> 4. In the logs I do see lots of DependenciesResolver Could not resolve assembly messages eg
>    [08:38:01.113 DBG] [PbiTools.Utils.DependenciesResolver] Could not resolve assembly: Microsoft.VisualStudio, PublicKeyToken=b03f5f7f11d50a3a
>    [08:38:02.959 DBG] [PbiTools.Utils.DependenciesResolver] Could not resolve assembly: Microsoft.AnalysisServices.Tabular.resources, Version=19.34.1.0, Culture=en-GB, PublicKeyToken=89845dcd8080cc91
>    [08:38:09.686 DBG] [PbiTools.Utils.DependenciesResolver] Could not resolve assembly: Microsoft.Mashup.Client.Packaging.XmlSerializers, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35

All of those can be safely ignored!

---
Interesting that the same issue occurs with a completely different file. Must be related to your system them. Is this a work machine that's somewhat locked down?

I also noticed you had extracted `pbi-tools` into C:\Program Files. That should generally be fine, however, may well be locked down on some systems. Could you try running it from another location, `C:\Temp` or so?

Thank you!

**Comment by TonyGaul** [comment]
Hi Mathias,

Thanks for the tips.  So I have followed along with tips and also set things up (folder structure, test file etc) as per this guide: (https://pbi.tools/tutorials/getting-started-cli.html).  Both my desktop and laptop are my own and are not running any corporate ware or restrictions.  I had to install .NET 6 from your link for `core`.

PBIX Testing
Before I talk about the test results, can you please confirm my expectations are correct?  My original PBIX is a single file solution with visuals and a data model, and I was expecting that I could `extract` the files to a structure then `compile` back into a PBIX with data model and visuals, and just refresh the model.  But after reading this (https://pbi.tools/cli/usage.html#compile) it seems I am expecting things that are not available yet?  "Generates a PBIX/PBIT file from sources in the specified PbixProj folder. Currently, the PBIX output is supported only for report-only projects ("thin" reports), and PBIT for projects containing a data model."  So right now I should not expect the compile to produce a PBIX with a refreshable data model, right?  If it is possible to do this, then I am still getting an error when opening a compile PBIX using :
`pbi-tools extract '.\Adventure Works DW 2020.pbix'`
`pbi-tools compile '.\Adventure Works DW 2020' -outPath AdvWorksDW2020.pbix`

***@***.***


`pbi-tools.core compile '.\Adventure Works DW 2020' -outPath AdvWorksDW2020.pbix -overwrite true` produces this error for me
[20:38:25.652 INF] [PbiTools.Model.PbixModel] Generating PBIX file at 'C:\pbi-tools-test\AdvWorksDW2020.pbix'...
[20:38:25.656 FTL] [] An unhandled exception occurred.
System.NotSupportedException: Compiling a project containing a data model into a PBIX file is not supported by the Core edition of pbi-tools.
   at PbiTools.Model.PbixModel.ToFile(String path, PbiFileFormat format, IDependenciesResolver dependenciesResolver) in X:\pbi-tools\BRANCHES\extract-watch\src\PBI-Tools\Model\PbixModel.cs:line 341
   at PbiTools.Cli.CmdLineActions.CompilePbix(String folder, String outPath, PbiFileFormat format, Boolean overwrite) in X:\pbi-tools\BRANCHES\extract-watch\src\PBI-Tools\Cli\Compile.cs:line 74
--- End of stack trace from previous location ---
   at PowerArgs.ArgAction.Invoke()
   at PbiTools.Program.Main(String[] args) in X:\pbi-tools\BRANCHES\extract-watch\src\PBI-Tools\Program.cs:line 84


PBIT Testing
`pbi-tools compile '.\Adventure Works DW 2020' -outPath AdvWorksDW2020.pbit -format PBIT` produces a good PBIT
`pbi-tools.core compile '.\Adventure Works DW 2020' -outPath AdvWorksDW2020.pbit -format PBIT` also produces a good PBIT

I appreciate anytime you can spare to help.  I might be an edge case, so understand if you have higher priorities right now.

Thanks,
Tony

**Comment by TonyGaul** [comment]
Hi Mathias,

Thanks for the tips.  So I have followed along with tips and also set things up (folder structure, test file etc) as per this guide: (https://pbi.tools/tutorials/getting-started-cli.html).  Both my desktop and laptop are my own and are not running any corporate ware or restrictions.  I had to install .NET 6 from your link for `core`.

**PBIX Testing**
Before I talk about the test results, can you please confirm my expectations are correct?  My original PBIX is a single file solution with visuals and a data model, and I was expecting that I could `extract` the files to a structure then `compile` back into a PBIX with data model and visuals, and just refresh the model.  But after reading this (https://pbi.tools/cli/usage.html#compile) it seems I am expecting things that are not available yet?  "Generates a PBIX/PBIT file from sources in the specified PbixProj folder. Currently, the PBIX output is supported only for report-only projects ("thin" reports), and PBIT for projects containing a data model."  So right now I should not expect the compile to produce a PBIX with a refreshable data model, right?  If it is possible to do this, then I am still getting an error when opening a compile PBIX using :
`pbi-tools extract '.\Adventure Works DW 2020.pbix'`
`pbi-tools compile '.\Adventure Works DW 2020' -outPath AdvWorksDW2020.pbix`

![image](https://user-images.githubusercontent.com/14303694/155840404-fccdcfdf-8ffa-4305-9fb4-9ae9a5873f00.png)



`pbi-tools.core compile '.\Adventure Works DW 2020' -outPath AdvWorksDW2020.pbix -overwrite true` produces this error for me
[20:38:25.652 INF] [PbiTools.Model.PbixModel] Generating PBIX file at 'C:\pbi-tools-test\AdvWorksDW2020.pbix'...
[20:38:25.656 FTL] [] An unhandled exception occurred.
System.NotSupportedException: Compiling a project containing a data model into a PBIX file is not supported by the Core edition of pbi-tools.
   at PbiTools.Model.PbixModel.ToFile(String path, PbiFileFormat format, IDependenciesResolver dependenciesResolver) in X:\pbi-tools\BRANCHES\extract-watch\src\PBI-Tools\Model\PbixModel.cs:line 341
   at PbiTools.Cli.CmdLineActions.CompilePbix(String folder, String outPath, PbiFileFormat format, Boolean overwrite) in X:\pbi-tools\BRANCHES\extract-watch\src\PBI-Tools\Cli\Compile.cs:line 74
--- End of stack trace from previous location ---
   at PowerArgs.ArgAction.Invoke()
   at PbiTools.Program.Main(String[] args) in X:\pbi-tools\BRANCHES\extract-watch\src\PBI-Tools\Program.cs:line 84


**PBIT Testing**
`pbi-tools compile '.\Adventure Works DW 2020' -outPath AdvWorksDW2020.pbit -format PBIT` produces a good PBIT
`pbi-tools.core compile '.\Adventure Works DW 2020' -outPath AdvWorksDW2020.pbit -format PBIT` also produces a good PBIT

I appreciate anytime you can spare to help.  I might be an edge case, so understand if you have higher priorities right now.

Thanks,
Tony

**Comment by mthierba** [comment]
Hi Tony,

glad I was able to clarify! And don't be sorry at all - those questions help me finding gaps in docs and comms.

Regarding model deployments. In Production, I'm currently using proprietary scripts deploying the outputs from `pbi-tools generate-bim` via XMLA endpoints.

It is on my agenda for March to fully bring those capabilities into `pbi-tools`: Deployment manifests will support a new "Dataset" mode so that the `compile` action can be used to deploy **Reports** to Power BI Pro/Premium/PPU as well as **Datasets** to AAS/PBI Premium/PPU.

I have been working on spec'ing the details in the background, and will track further progress on #97.

Possibilities already in place _today_:

- Deploy .PBIX containing full models using `pbi-tools deploy` and the "Report/File" mode. ***Until Microsoft changes their licensing strategy, that will always be the only option for Power BI Pro!***
- Use `pbi-tools generate-bim` against a PbixProj folder and deploy the resulting .BIM file using Tabular Editor or custom scripts.

**Comment by TonyGaul** [comment]
Great new, thanks Mathias.  Sounds like your baby is fast becoming a full time job.  Look for to using the new features.  Btw I am happy to test this out.  Would love to support the development, but not sure I have the necessary programming language skills to support!

**Comment by TonyGaul** [comment]
And I know you are a busy man, but if you ever happen to find time to go in depth on how you configure your DevOps, branching, feature releases that would be super appreciated.  I am lacking skills in that area, but am reading through your links in the pbi-tools.com site and will have go and building something myself.

Can I ask one more question with deployment, when you say check in one changed pbix into "dev" branch (one of many pbix's in the repo), that one pbix can be deployed file "file" mode.  Now when you move to UAT & Prod, can you control the deployment so only that one change pbix gets deployed to UAT & Prod, or does the entire project containing all the pbix files get deployed?

**Comment by mthierba** [comment]
> And I know you are a busy man, but if you ever happen to find time to go in depth on how you configure your DevOps, branching, feature releases that would be super appreciated. I am lacking skills in that area, but am reading through your links in the pbi-tools.com site and will have go and building something myself.

Stay tuned, something is going to come up on YouTube over the next few days... :wink:

> Can I ask one more question with deployment, when you say check in one changed pbix into "dev" branch (one of many pbix's in the repo), that one pbix can be deployed file "file" mode. Now when you move to UAT & Prod, can you control the deployment so only that one change pbix gets deployed to UAT & Prod, or does the entire project containing all the pbix files get deployed? 

The current deployment features in `pbi-tools` are unconditional: Each deployment run does re-deploy all artifacts contained in the respective git branch.

- For (thin) reports and XMLA datasets that approach is valid since those deployments are relatively "cheap", they are quick and unobtrusive.
- `pbi-tools` follows an important DevOps principle here: Deploying often and frequently ultimately makes the process more stable and reliable, and provides quick feedback when things go wrong.
- It also ensures that the git repo is conceptually and technically established as the Source-of-Truth, a very important design principle for me.
- Admittedly, for Power BI Pro deployments of datasets via fully-fledged .PBIX files that approach would not necessarily be desirable. Might need to look into a more varied approach. However, as of today this is also out of scope for me as I need to get the Dataset story developed for AAS and Premium first.

**Comment by TonyGaul** [comment]
Fantastic! I'll keep an eye out for that YouTube video. Look forward to it.

With the DevOps process / principals, I do agree. I currently use basic pipelines for ADF and dbt, and they do an entire project deployment. I guess with some of our reports, we have some UAT reports that have completed testing and some not finished testing yet. The tested ones sometimes need to be deployed to Prod, but the other ones not finished testing must not be.  This is not a big deal really. As you mention, the dataset deployment is much more important.  If I ever do need selective deployments, I might do a Dev pipeline, then use Power BI deployment pipelines to do a selective deployment of a subset reports and datasets from Dev->Uat->Prod . PBI deployment pipelines also allows me to maintain different param values in each environment. The big drawback is I am using 2 different deployment processes, and potentially environments are not in sync with source.

Anyway, I've taken enough of your time today! Thanks Tony

27 Feb 2022 9:38:18 pm Mathias Thierbach ***@***.***>:

> And I know you are a busy man, but if you ever happen to find time to go in depth on how you configure your DevOps, branching, feature releases that would be super appreciated. I am lacking skills in that area, but am reading through your links in the pbi-tools.com site and will have go and building something myself.
> 
> Stay tuned, something is going to come up on YouTube over the next few days... 😉
> 
> Can I ask one more question with deployment, when you say check in one changed pbix into "dev" branch (one of many pbix's in the repo), that one pbix can be deployed file "file" mode. Now when you move to UAT & Prod, can you control the deployment so only that one change pbix gets deployed to UAT & Prod, or does the entire project containing all the pbix files get deployed?
> 
> The current deployment features in *pbi-tools* are unconditional: Each deployment run does re-deploy all artifacts contained in the respective git branch.
> 
* > For (thin) reports and XMLA datasets that approach is valid since those deployments are relatively "cheap", they are quick and unobtrusive.
* > *pbi-tools* follows an important DevOps principle here: Deploying often and frequently ultimately makes the process more stable and reliable, and provides quick feedback when things go wrong.
* > It also ensures that the git repo is conceptually and technically established as the Source-of-Truth, a very important design principle for me.
* > Admittedly, for Power BI Pro deployments of datasets via fully-fledged .PBIX files that approach would not necessarily be desirable. Might need to look into a more varied approach. However, as of today this is also out of scope for me as I need to get the Dataset story developed for AAS and Premium first.
